core/devdoc/on-out-process-gateway-modules.md

```diff
@@ -0,0 +1,484 @@
+        },
+        "module.arguments" : {
+            "hostname": "<YourHostNameHere from Functions Portal>",
+            "relativePath": "api/<YourFunctionName>",
+            "key": "<your Api Key Here>"
+        }
+    }
+}
+]
+```
+
+Proxy/Module Host Control Communication Protocol
+------------------------------------------------
+
+<a name="control_communication" ></a>
+The proxy and the module host communicate via *messages* using *nanomsg*
+pair sockets. The different message types and how they work has been
+described below. The ordering of the message types given below corresponds
+roughly with the sequence in which these messages are sent and received. Each
+message packet is structured like so:
+
+![](./media/message-struct.png)
+
+-   **Create Module Request**
+
+    This message is sent by the proxy to the module host after it establishes a
+    socket connection. The data content will be as follows:
+    ```c
+    struct 
+        {
+        const uint32_t uri_type;
+        const uint32_t uri_size;
+        const uint8_t  message_channel_uri[uri_size];
+    } MESSAGE_URI;
+
+    const uint8_t gateway_message_version;
+    MESSAGE_URI  uri;
+    const uint32_t args_size;
+    const char     args[arg_size];
+    ```
+
+    where
+    >| Field                   | Description   |
+    >|-------------------------|---------------|
+    >| gateway_message_version | The version of gateway messages being sent |
+    >| uri_type                | Implementation specific URI type. (e.g NN_PAIR) |
+    >| message_channel_uri     | The message channel URI, for the socket which the proxy and outprocess module exchange gateway messages. |
+    >| uri_size                | Size of the messaging URI. |
+    >| args                    | Serialized JSON string of module arguments. |
+    >| args_size               | Size of the module arguments. |
+
+
+
+-   **Module Reply**
+
+    This message is sent by the module host in response to a *Create Module*
+    request from the proxy. The *data* in the packet is a boolean status
+    indicating whether the module creation was successful or not.
+
+-   **Start Module Request**
+
+    This message is sent by the proxy to the module host when `Module_Start` is called on the proxy module. 
+
+-   **Heartbeat Request**
+
+    Reserved for future use, this message will be sent by the proxy to the 
+    module host when it wants to confirm the module host is still running.
+
+-   **Heartbeat Reply**
+
+    Reserved for future use, this message will be sent by the module host to 
+    the proxy in repsonse to the heartbeat request. 
+
+-   **Destroy Module Request**
+
+    This message is sent by the proxy to the module host when the module is
+    being shut down. There is no *data* associated with this message and no
+    reply is expected to be sent by the module host. The module host is expected
+    to invoke `Module_Destroy` on the module and quit the process.
+
+Open Questions
+--------------
```

Comment:
Again, I know we discussed that before... but I would remove these open questions from here and translate that into PBI's (Or github projects). Having this here is yet another process to track new features IMO.